main/src/ca/uwaterloo/flix/language/ast/Purity.scala

```diff
@@ -114,19 +114,32 @@ object Purity {
   /**
     * Returns the purity of the given formula `eff`. Returns [[Pure]] for the
     * effect constant of [[TypeConstructor.Pure]], returns [[Impure]] for the
-    * effect constant of [[Symbol.IO]], and [[ControlImpure]] otherwise.
+    * effect constant of [[Symbol.IO]] or other Java effects, and [[ControlImpure]] otherwise.
     *
     * Assumes that the given type is a well-formed formula without variables,
     * aliases, or associated types.
     */
   def fromType(eff: Type)(implicit universe: Set[Symbol.EffectSym]): Purity = {
     evaluateFormula(eff) match {
       case set if set.isEmpty => Purity.Pure
-      case set if set.sizeIs == 1 && set.contains(Symbol.IO) => Purity.Impure
+      case set if set.forall(isJavaEffect) => Purity.Impure
       case _ => Purity.ControlImpure
     }
   }
 
+  def isJavaEffect(sym: Symbol.EffectSym): Boolean = sym match {
```

Comment:
Surely this should be somewhere else, fx right next to where these all are declared.

Reply:
@sockmaster27 Is adding this function also. You should coordinate and add it in one place.

Reply:
I propose adding it to `Symbol`:  #8699